File: apps/tvos/Packages/PlayerMPV/Sources/PlayerMPV/MPVErrorMapping.swift
// SPDX-FileCopyrightText: 2026 Spidola contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

import Libmpv
import PlayerContract

/// mpv's failure vocabulary translated into the contract's taxonomy (TECH_SPEC §8).
///
/// Pure by construction — every function here is a total function of its arguments, with no handle,
/// no I/O, and no clock. That is what lets the mapping be unit-tested branch by branch while the
/// decoder it describes cannot run on a build machine at all.
enum MPVErrorMapping {

  // MARK: - Log-line classification

  /// The failure a single mpv/FFmpeg log line implies, or `nil` if the line says nothing about why
  /// loading failed.
  ///
  /// This is a heuristic over human-readable strings, and it is worth being honest about why we are
  /// reduced to one: `MPV_ERROR_LOADING_FAILED` is mpv's single answer for "the URL did not open",
  /// covering DNS failure, a refused connection, an HTTP 401, and an unrecognised container alike.
  /// The distinction the PRD's error UX depends on — retry vs. fix your login vs. try the other
  /// player — exists only in the log text FFmpeg emitted on the way. So we read it.
  ///
  /// Matching is case-insensitive and substring-based rather than anchored, because these strings
  /// come from two upstreams (mpv's own messages and FFmpeg's protocol layer) and carry varying
  /// prefixes. A miss returns `nil` and degrades to the code-based mapping — never to a wrong
  /// class, which would send the viewer down the wrong recovery path.
  static func logHint(from text: String) -> EngineError? {
    let line = text.lowercased()

    // Auth first: an HTTP 401/403 is also reported as a generic "failed to open", so a later
    // unreachable-shaped match must not win over it.
    if line.contains("401") && line.contains("unauthorized") { return .unauthorized }
    if line.contains("403") && line.contains("forbidden") { return .unauthorized }
    if line.contains("server returned 401") || line.contains("server returned 403") {
      return .unauthorized
    }
    if line.contains("authorization failed") || line.contains("authentication failed") {
      return .unauthorized
    }

    // Reachability: name resolution, refused/unreachable peers, and HTTP 404/5xx all mean the same
    // thing to the viewer — the channel's server did not answer usefully — and no engine swap
    // would change that.
    if line.contains("failed to resolve") || line.contains("name or service not known") {
      return .sourceUnreachable
    }
    if line.contains("connection refused") || line.contains("connection reset") {
      return .sourceUnreachable
    }
    if line.contains("network is unreachable") || line.contains("no route to host") {
      return .sourceUnreachable
    }
    if line.contains("server returned 404") || line.contains("server returned 5") {
      return .sourceUnreachable
    }
    if line.contains("tcp_connect") || line.contains("connection timed out") {
      return .sourceUnreachable
    }

    // Demux: the bytes arrived but nothing could read them as a container.
    if line.contains("failed to recognize file format") { return .unsupportedFormat }
    if line.contains("could not determine file format") { return .unsupportedFormat }
    if line.contains("invalid data found when processing input") { return .unsupportedFormat }
    if line.contains("no such file or directory") { return .unsupportedFormat }

    // Decode: the container opened and a codec inside it failed.
    if line.contains("could not open codec") || line.contains("no decoder found") {
      return .decoderFailed
    }
    if line.contains("decoder init failed") || line.contains("could not initialize video chain") {
      return .decoderFailed
    }
    if line.contains("could not open video decoder")
      || line.contains("could not open audio decoder")
    {
      return .decoderFailed
    }

    return nil
  }

  // MARK: - Error-code mapping

  /// The failure an mpv error code implies, refined by `logHint` where the code alone is ambiguous.
  ///
  /// `logHint` is consulted only for `MPV_ERROR_LOADING_FAILED`. The other codes are unambiguous on
  /// their own, and letting a stray log line override them would make the mapping depend on which
  /// message happened to be logged last — the exact non-determinism this function exists to avoid.
  static func engineError(mpvError: Int32, logHint: EngineError? = nil) -> EngineError {
    switch mpvError {
    case MPV_ERROR_LOADING_FAILED.rawValue:
      // The disambiguation point. Absent a hint we say unreachable: for a stream URL that would
      // not open, "the server didn't answer" is both the likeliest cause and the honest one — it
      // offers retry rather than the "Try other player" button, which cannot help here and would
      // waste the viewer's time (EngineError.offersOtherPlayer).
      return logHint ?? .sourceUnreachable

    case MPV_ERROR_UNKNOWN_FORMAT.rawValue, MPV_ERROR_NOTHING_TO_PLAY.rawValue:
      // Nothing demuxable in what arrived — the one case where another engine genuinely might win.
      return .unsupportedFormat

    case MPV_ERROR_UNSUPPORTED.rawValue:
      return .unsupportedFormat

    case MPV_ERROR_AO_INIT_FAILED.rawValue, MPV_ERROR_VO_INIT_FAILED.rawValue:
      // The stream was fine; our own output chain would not start. Classed as a decode failure
      // because that is the class whose recovery ("try the other player") actually applies.
      return .decoderFailed

    case MPV_ERROR_NOMEM.rawValue:
      return .decoderFailed

    case MPV_ERROR_SUCCESS.rawValue:
      // Reached only if a caller asks to classify a non-failure. Not representable as "no error"
      // because the return type is non-optional by design: every call site here is already on a
      // failure path, so a success code is a bug in the caller, not a state to model.
      return .unknown(detail: "mpv reported success on a failure path")

    default:
      return .unknown(detail: describe(mpvError: mpvError))
    }
  }

  // MARK: - End-of-file mapping

  /// The contract state an mpv `MPV_EVENT_END_FILE` implies, or `nil` when the end-of-file is one
  /// we caused and the shell should not be told anything.
  ///
  /// `STOP` and `QUIT` are returned as `nil` deliberately: both are the fingerprints of our own
  /// teardown, and reporting them would race a `.failed` or `.ended` onto the stream while the
  /// engine is being disposed on a zap — turning every channel change into a spurious error.
  /// `REDIRECT` is likewise silent; mpv is about to load the real target and the state machine
  /// should stay in `.loading` across it.
  static func endFileOutcome(
    reason: mpv_end_file_reason,
    mpvError: Int32,
    logHint: EngineError? = nil
  ) -> PlaybackState? {
    switch reason {
    case MPV_END_FILE_REASON_EOF:
      return .ended
    case MPV_END_FILE_REASON_ERROR:
      return .failed(engineError(mpvError: mpvError, logHint: logHint))
    case MPV_END_FILE_REASON_STOP, MPV_END_FILE_REASON_QUIT, MPV_END_FILE_REASON_REDIRECT:
      return nil
    default:
      // mpv's reason list has grown before and may grow again; an unrecognised reason is not a
      // failure to report at the viewer.
      return nil
    }
  }

  // MARK: - Diagnostics

  /// A safe-by-construction description of an mpv error code.
  ///
  /// `mpv_error_string` returns a pointer into a static table compiled into libmpv — a fixed
  /// English phrase like "loading failed". It cannot contain stream data, so unlike mpv's log
  /// text it is safe to put in `EngineError.unknown(detail:)`, which reaches the log stream
  /// (TECH_SPEC §12).
  static func describe(mpvError: Int32) -> String {
    guard let cString = mpv_error_string(mpvError) else {
      return "mpv error \(mpvError)"
    }
    return "mpv: \(String(cString: cString)) (\(mpvError))"
  }
}

File: apps/androidtv/player/engine-mpv/src/main/kotlin/dev/spidola/tv/player/enginempv/MpvErrorMapping.kt
// SPDX-FileCopyrightText: 2026 Spidola contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

package dev.spidola.tv.player.enginempv

import dev.spidola.tv.core.playercontract.EngineError

/**
 * mpv's failure vocabulary → the contract's [EngineError] taxonomy (TECH_SPEC §8).
 *
 * Pure, and deliberately free of any native reference so it is unit-testable on the JVM with
 * no `libmpv.so` present — which matters, because this mapping decides whether the viewer is
 * offered "Try other player", and that decision must be tested over every branch rather than
 * on a device with one broken stream.
 *
 * **Why the log text is an input.** mpv's own error codes cannot distinguish the cases the
 * PRD's error UX must distinguish. Every one of "DNS failed", "connection refused",
 * "HTTP 401", and "HTTP 403" arrives as the single code `MPV_ERROR_LOADING_FAILED` (-13).
 * The only place that distinction survives is FFmpeg's log line. So classification consults
 * the diagnostic text first and the code second — not because text-matching is pleasant, but
 * because the alternative is collapsing four different viewer-facing outcomes into one
 * unhelpful "couldn't play".
 *
 * The diagnostic passed in **must already be redacted** ([MpvLogRedaction]); this file puts
 * it into [EngineError.Unknown.detail], which reaches the log stream.
 */
internal object MpvErrorMapping {
    /**
     * mpv's error codes, from `client.h`. Only the ones this mapping actually decides on are
     * named — an unnamed code lands in the `else` arm and becomes [EngineError.Unknown],
     * which is the honest answer for a code we have not reasoned about.
     */
    object Code {
        const val SUCCESS = 0
        const val LOADING_FAILED = -13
        const val AO_INIT_FAILED = -14
        const val VO_INIT_FAILED = -15
        const val NOTHING_TO_PLAY = -16
        const val UNKNOWN_FORMAT = -17
        const val UNSUPPORTED = -18
        const val GENERIC = -20
    }

    /** `MPV_END_FILE_REASON_*`, from `client.h`. */
    object EndFileReason {
        const val EOF = 0
        const val STOP = 2
        const val QUIT = 3
        const val ERROR = 4
        const val REDIRECT = 5
    }

    // Matched against the redacted diagnostic. Ordered most-specific-first by the caller.
    private val UNAUTHORIZED = Regex("""\b(401|403)\b|\bunauthorized\b|\bforbidden\b""", RegexOption.IGNORE_CASE)
    private val UNREACHABLE =
        Regex(
            """\b(name or service not known|temporary failure in name resolution|no such host|""" +
                """could not resolve|failed to resolve|connection refused|network is unreachable|""" +
                """no route to host|host is unreachable|connection reset)\b""",
            RegexOption.IGNORE_CASE,
        )
    private val TIMEOUT =
        Regex("""\b(timed out|timeout|connection timed out|operation timed out)\b""", RegexOption.IGNORE_CASE)
    private val UNSUPPORTED_FORMAT =
        Regex(
            """\b(invalid data found when processing input|unknown format|could not (find|open) codec|""" +
                """no decoder for|demux(ing|er)? fail|failed to (recognize|open) file format|""" +
                """unsupported|could not demux)\b""",
            RegexOption.IGNORE_CASE,
        )
    private val DECODE_FAILED =
        Regex(
            """\b(error while decoding|decoder error|failed to decode|hardware decoding failed|""" +
                """could not initialize (video|audio) (chain|decoder)|error initializing decoder|""" +
                """decode_slice_header error)\b""",
            RegexOption.IGNORE_CASE,
        )

    /**
     * Classifies an `MPV_EVENT_END_FILE`.
     *
     * Returns `null` when the file ended for a reason that is not a failure, so the caller
     * does not have to re-derive "was this actually bad": `EOF` is the stream running out
     * (the contract's `Ended`), and `STOP`/`QUIT`/`REDIRECT` are our own teardown or mpv
     * following a redirect — none of which the viewer should ever see an error for.
     */
    fun endFileError(
        reason: Int,
        errorCode: Int,
        diagnostic: String?,
    ): EngineError? =
        when (reason) {
            EndFileReason.ERROR -> engineErrorFor(errorCode, diagnostic)
            EndFileReason.EOF ->
                classifyDiagnostic(diagnostic)
                    ?.takeIf { it == EngineError.DecoderFailed }
            EndFileReason.STOP, EndFileReason.QUIT, EndFileReason.REDIRECT -> null
            // A reason mpv grew after this was written. Treating it as a failure would invent
            // an error the viewer cannot act on; ignoring it keeps the state machine honest.
            else -> null
        }

    /**
     * Classifies an mpv error code plus whatever mpv said about it.
     *
     * [diagnostic] is consulted before [errorCode] because it is strictly more specific: the
     * code says "loading failed", the text says why.
     */
    fun engineErrorFor(
        errorCode: Int,
        diagnostic: String?,
    ): EngineError {
        classifyDiagnostic(diagnostic)?.let { return it }

        return when (errorCode) {
            // The container or protocol was never openable.
            Code.UNKNOWN_FORMAT, Code.NOTHING_TO_PLAY, Code.UNSUPPORTED -> EngineError.UnsupportedFormat
            // The chain came up but a decoder or output could not initialize. Offering the
            // other engine here is honest: a different decoder genuinely might succeed.
            Code.VO_INIT_FAILED, Code.AO_INIT_FAILED -> EngineError.DecoderFailed
            // -13 with nothing quotable in the log. It is overwhelmingly a network failure,
            // but "overwhelmingly" is not "certainly", and guessing SourceUnreachable here
            // would suppress the "Try other player" button for a stream another engine could
            // have played. Unknown keeps the guess out of the viewer's way.
            Code.LOADING_FAILED -> EngineError.Unknown(detail(errorCode, diagnostic))
            Code.GENERIC -> EngineError.Unknown(detail(errorCode, diagnostic))
            else -> EngineError.Unknown(detail(errorCode, diagnostic))
        }
    }

    private fun classifyDiagnostic(diagnostic: String?): EngineError? {
        val text = diagnostic?.takeIf { it.isNotBlank() } ?: return null
        return when {
            // Auth before reachability: a 401 body arrives over a connection that plainly
            // worked, so a "connection reset" later in the same log must not outrank it.
            UNAUTHORIZED.containsMatchIn(text) -> EngineError.Unauthorized
            TIMEOUT.containsMatchIn(text) -> EngineError.Timeout
            UNREACHABLE.containsMatchIn(text) -> EngineError.SourceUnreachable
            // Decode before format: "error while decoding" implies the demuxer already
            // succeeded, so it is the more specific claim.
            DECODE_FAILED.containsMatchIn(text) -> EngineError.DecoderFailed
            UNSUPPORTED_FORMAT.containsMatchIn(text) -> EngineError.UnsupportedFormat
            else -> null
        }
    }

    /** Diagnostic text for the log stream (PRD §8.6: never for the screen). */
    private fun detail(
        errorCode: Int,
        diagnostic: String?,
    ): String =
        diagnostic
            ?.takeIf { it.isNotBlank() }
            ?.let { "mpv error $errorCode: $it" }
            ?: "mpv error $errorCode"
}
